File: src/db/migrate.js
import { readdirSync, readFileSync } from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { sql, openPool, request } from './pool.js';
import { sha256Text } from '../util/hash.js';
import { logger } from '../util/logger.js';

/**
 * Migration runner.
 *
 * Production uses these same files with an externally supplied connection string; tests
 * do not substitute SQLite or another engine, because a constraint that only exists in
 * one dialect is a constraint that was never tested.
 */

// The SQL migrations are a language-neutral asset and now live with the Python package
// during the port. This file is the superseded JavaScript reference implementation; it
// reads the same single canonical directory so both implementations stay in step and the
// parity comparison is against identical schema.
const MIGRATIONS_DIR = path.resolve(
  path.dirname(fileURLToPath(import.meta.url)),
  '..',
  'alerts_bi',
  'db',
  'migrations',
);

/**
 * @typedef {object} MigrationFile
 * @property {string} version
 * @property {string} filename
 * @property {string} sqlText
 * @property {string} checksum
 */

/**
 * Read migrations from disk in version order.
 * @returns {MigrationFile[]}
 */
export function loadMigrations() {
  return readdirSync(MIGRATIONS_DIR)
    .filter((f) => f.endsWith('.sql'))
    .sort()
    .map((filename) => {
      const sqlText = readFileSync(path.join(MIGRATIONS_DIR, filename), 'utf8');
      return {
        version: filename.replace(/\.sql$/, ''),
        filename,
        sqlText,
        checksum: sha256Text(sqlText.replace(/\r\n/g, '\n')),
      };
    });
}

/**
 * @param {any} pool
 * @returns {Promise<void>}
 */
async function ensureMigrationsTable(pool) {
  await request(pool).query(`
    IF OBJECT_ID('schema_migrations', 'U') IS NULL
    CREATE TABLE schema_migrations (
      version    NVARCHAR(128) NOT NULL CONSTRAINT pk_schema_migrations PRIMARY KEY,
      checksum   NVARCHAR(64)  NOT NULL,
      applied_at DATETIME2(3)  NOT NULL
    );
  `);
}

/**
 * @param {any} pool
 * @returns {Promise<Map<string, string>>} version -> checksum
 */
export async function appliedMigrations(pool) {
  await ensureMigrationsTable(pool);
  const result = await request(pool).query('SELECT version, checksum FROM schema_migrations');
  return new Map(result.recordset.map((r) => [r.version, r.checksum]));
}

/**
 * Apply every migration not yet recorded.
 *
 * A migration whose file changed after being applied is a hard error: silently
 * re-applying or ignoring it would leave the database in a state no version describes.
 *
 * @param {any} pool
 * @returns {Promise<{applied: string[], alreadyApplied: string[]}>}
 */
export async function migrate(pool) {
  const applied = await appliedMigrations(pool);
  const files = loadMigrations();

  /** @type {string[]} */
  const newlyApplied = [];
  /** @type {string[]} */
  const alreadyApplied = [];

  for (const migration of files) {
    const existing = applied.get(migration.version);
    if (existing !== undefined) {
      if (existing !== migration.checksum) {
        throw new Error(
          `migration ${migration.filename} was already applied but its contents have changed ` +
            '(add a new migration instead of editing an applied one)',
        );
      }
      alreadyApplied.push(migration.version);
      continue;
    }

    // Each migration runs in its own transaction: a failure leaves the previous ones
    // applied and recorded, so a rerun resumes rather than starting over.
    const transaction = new sql.Transaction(pool);
    await transaction.begin();
    try {
      // GO is a batch separator understood by sqlcmd, not by the driver.
      for (const batch of migration.sqlText.split(/^\s*GO\s*$/im)) {
        if (batch.trim() === '') continue;
        await request(transaction).batch(batch);
      }
      await request(transaction, {
        version: { type: sql.NVarChar(128), value: migration.version },
        checksum: { type: sql.NVarChar(64), value: migration.checksum },
        appliedAt: { type: sql.DateTime2, value: new Date() },
      }).query(
        'INSERT INTO schema_migrations (version, checksum, applied_at) VALUES (@version, @checksum, @appliedAt)',
      );
      await transaction.commit();
      newlyApplied.push(migration.version);
      logger.info('db.migration_applied', { version: migration.version });
    } catch (err) {
      await transaction.rollback();
      throw new Error(
        `migration ${migration.filename} failed: ${/** @type {any} */ (err).message}`,
      );
    }
  }

  return { applied: newlyApplied, alreadyApplied };
}

/**
 * Create the target database if it does not exist, then apply migrations to it.
 *
 * @param {import('../config/env.js').SqlConfig} config
 * @param {string} database
 * @returns {Promise<{applied: string[], alreadyApplied: string[]}>}
 */
export async function migrateDatabase(config, database) {
  const master = await openPool(config, 'master');
  try {
    const exists = await request(master, {
      name: { type: sql.NVarChar(128), value: database },
    }).query('SELECT 1 AS present FROM sys.databases WHERE name = @name');
    if (exists.recordset.length === 0) {
      // The database name is validated before reaching here; sys.databases cannot be
      // queried with a parameterized CREATE, so the identifier is quoted defensively.
      await request(master).batch(`CREATE DATABASE ${quoteIdentifier(database)}`);
      logger.info('db.database_created', { database });
    }
  } finally {
    await master.close();
  }

  const pool = await openPool(config, database);
  try {
    return await migrate(pool);
  } finally {
    await pool.close();
  }
}

/**
 * Quote a SQL Server identifier, rejecting anything that is not a plain name.
 *
 * Database names reach this from configuration, and the drop path below is destructive,
 * so the safe set is deliberately narrow.
 *
 * @param {string} name
 * @returns {string}
 */
export function quoteIdentifier(name) {
  if (!/^[A-Za-z_][A-Za-z0-9_]{0,127}$/.test(name)) {
    throw new Error(`unsafe SQL identifier: ${JSON.stringify(name)}`);
  }
  return `[${name}]`;
}

/**
 * Drop and recreate ONLY the explicitly named disposable test database.
 *
 * Guarded on purpose: this is the one destructive operation in the codebase, and the
 * repository instructions require verifying the target before recreating anything. It
 * refuses any database whose name is not the configured test database, so a mistyped
 * environment variable cannot take out `alerts_bi_dev` or a production store.
 *
 * @param {import('../config/env.js').SqlConfig} config
 * @param {string} database
 * @returns {Promise<void>}
 */
export async function resetTestDatabase(config, database) {
  if (database !== config.testDatabase) {
    throw new Error(
      `refusing to reset ${JSON.stringify(database)}: only the configured test database ` +
        `(${JSON.stringify(config.testDatabase)}) may be recreated`,
    );
  }
  if (!/test/i.test(database)) {
    throw new Error(
      `refusing to reset ${JSON.stringify(database)}: a disposable test database name must contain "test"`,
    );
  }

  const master = await openPool(config, 'master');
  try {
    const quoted = quoteIdentifier(database);
    // Single-user mode rolls back open sessions so the drop cannot hang behind a
    // connection left over from an interrupted test run.
    await request(master).batch(`
      IF EXISTS (SELECT 1 FROM sys.databases WHERE name = N'${database}')
      BEGIN
        ALTER DATABASE ${quoted} SET SINGLE_USER WITH ROLLBACK IMMEDIATE;
        DROP DATABASE ${quoted};
      END
    `);
    await request(master).batch(`CREATE DATABASE ${quoted}`);
    logger.info('db.test_database_recreated', { database });
  } finally {
    await master.close();
  }

  const pool = await openPool(config, database);
  try {
    await migrate(pool);
  } finally {
    await pool.close();
  }
}

File: src/db/migrate.py
"""Migration runner.

Production uses these same files with an externally supplied connection string; tests do
not substitute SQLite or another engine, because a constraint that only exists in one
dialect is a constraint that was never tested.

The migration files are unchanged from the JavaScript implementation, and the checksum is
computed the same way (line endings normalized to LF before hashing), so a database
migrated by either implementation is accepted by the other.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import UTC, datetime
from pathlib import Path

from src.config import SqlConfig
from src.db.connection import Database, connect, quote_identifier
from src.hashing import sha256_text
from src.logging_setup import log

__all__ = [
    "MIGRATIONS_DIR",
    "Migration",
    "applied_migrations",
    "load_migrations",
    "migrate",
    "migrate_database",
    "reset_test_database",
]

MIGRATIONS_DIR = Path(__file__).resolve().parent / "migrations"

#: ``GO`` is a batch separator understood by sqlcmd, not by the driver.
_GO = re.compile(r"^\s*GO\s*$", re.MULTILINE | re.IGNORECASE)


@dataclass(frozen=True, slots=True)
class Migration:
    version: str
    filename: str
    sql_text: str
    checksum: str


def load_migrations(directory: Path = MIGRATIONS_DIR) -> list[Migration]:
    """Read migrations from disk in version order."""
    migrations: list[Migration] = []
    for path in sorted(directory.glob("*.sql")):
        sql_text = path.read_bytes().decode("utf-8")
        migrations.append(
            Migration(
                version=path.stem,
                filename=path.name,
                sql_text=sql_text,
                checksum=sha256_text(sql_text.replace("\r\n", "\n")),
            )
        )
    return migrations


def _ensure_migrations_table(db: Database) -> None:
    db.execute(
        """
        IF OBJECT_ID('schema_migrations', 'U') IS NULL
        CREATE TABLE schema_migrations (
          version    NVARCHAR(128) NOT NULL CONSTRAINT pk_schema_migrations PRIMARY KEY,
          checksum   NVARCHAR(64)  NOT NULL,
          applied_at DATETIME2(3)  NOT NULL
        );
        """
    )
    db.commit()


def applied_migrations(db: Database) -> dict[str, str]:
    """Return ``{version: checksum}`` for migrations already applied."""
    _ensure_migrations_table(db)
    rows = db.query("SELECT version, checksum FROM schema_migrations")
    return {str(row["version"]): str(row["checksum"]) for row in rows}


def migrate(db: Database, directory: Path = MIGRATIONS_DIR) -> tuple[list[str], list[str]]:
    """Apply every migration not yet recorded.

    A migration whose file changed after being applied is a hard error: silently
    re-applying or ignoring it would leave the database in a state no version describes.

    Returns ``(newly_applied, already_applied)``.
    """
    applied = applied_migrations(db)
    newly_applied: list[str] = []
    already_applied: list[str] = []

    for migration in load_migrations(directory):
        existing = applied.get(migration.version)
        if existing is not None:
            if existing != migration.checksum:
                raise RuntimeError(
                    f"migration {migration.filename} was already applied but its contents "
                    "have changed (add a new migration instead of editing an applied one)"
                )
            already_applied.append(migration.version)
            continue

        # Each migration runs in its own transaction: a failure leaves the previous ones
        # applied and recorded, so a rerun resumes rather than starting over.
        try:
            for batch in _GO.split(migration.sql_text):
                if batch.strip():
                    db.execute(batch)
            db.execute(
                "INSERT INTO schema_migrations (version, checksum, applied_at) "
                "VALUES (:version, :checksum, :applied_at)",
                {
                    "version": migration.version,
                    "checksum": migration.checksum,
                    "applied_at": datetime.now(UTC).replace(tzinfo=None),
                },
            )
            db.commit()
        except Exception as exc:
            db.rollback()
            raise RuntimeError(f"migration {migration.filename} failed: {exc}") from exc

        newly_applied.append(migration.version)
        log.info("db.migration_applied", version=migration.version)

    return newly_applied, already_applied


def migrate_database(config: SqlConfig, database: str) -> tuple[list[str], list[str]]:
    """Create the target database if it does not exist, then apply migrations to it."""
    quoted = quote_identifier(database)
    with connect(config, "master", autocommit=True) as master:
        exists = master.query_one(
            "SELECT 1 AS present FROM sys.databases WHERE name = :name", {"name": database}
        )
        if exists is None:
            master.execute(f"CREATE DATABASE {quoted}")
            log.info("db.database_created", database=database)

    with connect(config, database) as db:
        return migrate(db)


def reset_test_database(config: SqlConfig, database: str) -> None:
    """Drop and recreate ONLY the explicitly named disposable test database.

    Guarded on purpose: this is the one destructive operation in the codebase. It refuses
    any database whose name is not the configured test database, so a mistyped environment
    variable cannot take out ``alerts_bi_dev`` or a production store.
    """
    if database != config.test_database:
        raise ValueError(
            f"refusing to reset {database!r}: only the configured test database "
            f"({config.test_database!r}) may be recreated"
        )
    if "test" not in database.lower():
        raise ValueError(
            f'refusing to reset {database!r}: a disposable test database name must contain "test"'
        )

    quoted = quote_identifier(database)
    with connect(config, "master", autocommit=True) as master:
        # Single-user mode rolls back open sessions so the drop cannot hang behind a
        # connection left over from an interrupted test run.
        master.execute(
            f"""
            IF EXISTS (SELECT 1 FROM sys.databases WHERE name = N'{database}')
            BEGIN
              ALTER DATABASE {quoted} SET SINGLE_USER WITH ROLLBACK IMMEDIATE;
              DROP DATABASE {quoted};
            END
            """
        )
        master.execute(f"CREATE DATABASE {quoted}")
        log.info("db.test_database_recreated", database=database)

    with connect(config, database) as db:
        migrate(db)
